pkg/client/client.go

```diff
@@ -1298,27 +1299,47 @@ func (c *Client) WaitForDaemonSetRollout(ctx context.Context, ds *appsv1.DaemonS
 			klog.V(4).ErrorS(err, "WaitForDaemonSetRollout: failed to get DaemonSet")
 			return false, nil
 		}
+		want := d.Status.DesiredNumberScheduled
+		have := d.Status.NumberAvailable
+		numberUnavailable := want - have
+		maxUnavailableIntStr := intstr.FromInt(1)
+
+		if d.Spec.UpdateStrategy.RollingUpdate != nil && d.Spec.UpdateStrategy.RollingUpdate.MaxUnavailable != nil {
+			maxUnavailableIntStr = *d.Spec.UpdateStrategy.RollingUpdate.MaxUnavailable
+		}
+		maxUnavailable, intstrErr := intstr.GetScaledValueFromIntOrPercent(&maxUnavailableIntStr, int(want), true)
+
+		if intstrErr != nil {
+			lastErr = errors.Errorf("The daemonset has an invalid MaxUnavailable value: %v", intstrErr)
+			return false, nil
+		}
+
+		if int(numberUnavailable) > maxUnavailable {
+			lastErr = errors.Errorf("Too many daemonset pods are unavailable (%d > %d max unavailable).", numberUnavailable, maxUnavailable)
+			return false, nil
+		}
+
 		if d.Generation > d.Status.ObservedGeneration {
 			lastErr = errors.Errorf("current generation %d, observed generation: %d",
 				d.Generation, d.Status.ObservedGeneration)
 			return false, nil
 		}
```

Comment:
this should always the first check: it guarantees that the reported status is up-to-date.

Reply:
moved it as first case